compiler/z/codegen/OMRTreeEvaluator.cpp

```diff
@@ -2004,49 +2004,41 @@ genLoadAddressConstant(TR::CodeGenerator * cg, TR::Node * node, uintptr_t value,
       return generateRegLitRefInstruction(cg, TR::InstOpCode::getLoadOpCode(), node, targetRegister, (uintptr_t) value, reloKind, cond, cursor, base);
       }
 
-   TR::Symbol *symbol = NULL;
-   TR::Compilation *comp = cg->comp();
+   if (node->getOpCode().hasSymbolReference())
+      return genLoadLongConstant(cg, node, (uint64_t) value, targetRegister, cursor, cond, base);
 
-   if (node && node->getOpCode().hasSymbolReference())
-      symbol = node->getSymbol();
-   bool isPICCandidate = symbol ? symbol->isStatic() && symbol->isClassObject() : false;
-   if (isPICCandidate && !cg->comp()->compileRelocatableCode()
-       && cg->wantToPatchClassPointer((TR_OpaqueClassBlock*)value, node))
-      return genLoadAddressConstantInSnippet(cg, node, (uintptr_t)value, targetRegister, cursor, cond, base, true);
+   cursor = generateRILInstruction(cg, TR::InstOpCode::LLILF, node, targetRegister, static_cast<int32_t>(value), cursor);
 
+   TR::Compilation *comp = cg->comp();
+   bool assumePatch = false;
+   bool isCompressedClassPointer = false;
+   bool over32Bits = (value & CONSTANT64(0xFFFFFFFF00000000)) != 0;
    if (node->isClassUnloadingConst())
       {
-      uintptr_t value = node->getAddress();
-      TR::Instruction *unloadableConstInstr = NULL;
-      if (cg->canUseRelativeLongInstructions(value))
-         {
-         unloadableConstInstr = generateRILInstruction(cg, TR::InstOpCode::LARL, node, targetRegister, reinterpret_cast<void*>(value));
-         }
-      else
-         {
-         unloadableConstInstr = genLoadAddressConstantInSnippet(cg, node, value, targetRegister, NULL, NULL, NULL, true);
-         }
-
       if (node->isMethodPointerConstant())
          {
-         comp->getStaticMethodPICSites()->push_front(unloadableConstInstr);
+         comp->getStaticMethodPICSites()->push_front(cursor);
          }
       else
          {
-         comp->getStaticPICSites()->push_front(unloadableConstInstr);
+         comp->getStaticPICSites()->push_front(cursor);
+         isCompressedClassPointer = comp->useCompressedPointers();
          }
-      return unloadableConstInstr;
+      assumePatch = true;
       }
-
-   if (cg->comp()->target().is64Bit())
+   if (comp->getOption(TR_EnableHCR))
       {
-      return genLoadLongConstant(cg, node, (uint64_t) value, targetRegister, cursor, cond, base);
+      comp->getStaticHCRPICSites()->push_front(cursor);
+      assumePatch = true;
       }
-   else
+
+   if (comp->target().is64Bit() && !isCompressedClassPointer && (assumePatch || over32Bits))
```

Comment:
I'm assuming here that a class pointer in compressed refs mode will always fit in 32 bits. @0xdaryl, I've been told you are looking at changing that on Aarch64 MacOSX. Are those changes specific to that target, or will they affect my assumption here eventually?

Reply:
Can we place a comment here for the taken assumption? Also I would wait for what @0xdaryl  says, but if that the assumption is that class pointer constants are 32 bit for compressed refs, least we should have an ASSERT for that.